packages/react-compose/src/compose.ts

```diff
@@ -1,37 +1,66 @@
 import * as React from 'react';
+import {
+  ComponentWithAs,
+  ComposedComponent,
+  ComposeOptions,
+  Input,
+  InputComposeComponent,
+  ComposePreparedOptions,
+} from './types';
+import { mergeComposeOptions } from './mergeComposeOptions';
 
-import { ComponentWithAs, ComposedComponent, ComposeOptions, Input, InputComposeComponent } from './types';
-import { mergeComposeOptions, wasComposedPreviously } from './utils';
-
-function compose<ElementType extends React.ElementType, InputProps, InputStylesProps, ParentProps, ParentStylesProps>(
-  input: Input<ElementType, InputProps>,
-  inputOptions: ComposeOptions<InputProps, InputStylesProps, ParentProps, ParentStylesProps> = {},
+/**
+ * The `compose` utility allows React components to be created using a set of default config settings. Resulting
+ * component can be "recomposed" by calling `compose` with them as the first argument, attaching config overrides
+ * in the second argument. Settings are merged correctly and provided to the component render function.
+ *
+ * @param input - render function `(props, ref, config) => JSX.Element` or previously composed component.
+ * @param inputOptions - options for rendering the component.
+ */
+function compose<
+  TElementType extends React.ElementType,
+  TProps = {},
+  TInputStylesProps = {},
+  TParentProps = {},
+  TParentStylesProps = {},
+  TState = TProps
+>(
+  input: Input<TElementType, TProps>,
+  inputOptions: ComposeOptions<TProps, TInputStylesProps, TParentProps, TParentStylesProps> = {},
 ) {
-  const composeOptions = mergeComposeOptions(
+  // Merge the options as needed.
+  const composeOptions = mergeComposeOptions<TElementType, TProps, TState>(
     input as Input,
     (inputOptions as unknown) as ComposeOptions,
-    wasComposedPreviously(input) ? input.fluentComposeConfig : undefined,
+    // tslint:disable-next-line: no-any
+    (input as any).fluentComposeConfig,
   );
 
-  const Component = (React.forwardRef<ElementType, InputProps & ParentProps & { as?: React.ElementType }>(
+  // Create the new component, passing along the options.
+  const Component = (React.forwardRef<TElementType, TProps & TParentProps & { as?: React.ElementType }>(
     (props, ref) => {
-      return composeOptions.render(props, (ref as unknown) as React.Ref<HTMLDivElement>, {
-        ...composeOptions,
-        slots: {
-          ...composeOptions.slots,
-          __self: Component,
-        },
-      });
+      // tslint:disable-next-line:no-any
+      return composeOptions.render(props, ref as any, (composeOptions as unknown) as ComposePreparedOptions);
+      // return composeOptions.render(props, (ref as unknown) as React.Ref<HTMLDivElement>, {
+      //   ...composeOptions,
+      //   slots: {
+      //     ...composeOptions.slots,
+      //     __self: Component,
```

Comment:
We need to preserve the `__self` it us used at least in the Menu component

Reply:
got it in my reworked pr